Compactor: fix a race condition between different compactor replicas causing "corrupted" bucket index

pracucci opened the pull request:
#### What this PR does

This PR should fix #8687. The issue describes the problem, and the simple idea of this solution is explained in the (intentionally) long comment in `pkg/storage/tsdb/bucketindex/updater.go`. The test I've added fails on `main` and passes in this PR.

#### Which issue(s) this PR fixes or relates to

Fixes #8687

#### Checklist

- [x] Tests updated.
- [ ] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

dimitarvdimitrov reviewed (approved):
LGTM. I still haven't figured out how to have a more general test for block cleaning concurrency. Thanks for the fix

pracucci commented:
> I still haven't figured out how to have a more general test for block cleaning concurrency. Thanks for the fix

Me neither. I tried reproducing the issue adding random sleeps but I couldn't reproduce it.

seizethedave commented:
> >I still haven't figured out how to have a more general test for block cleaning concurrency. Thanks for the fix
> 
> Me neither. I tried reproducing the issue adding random sleeps in the objstore.Bucket functions but I couldn't reproduce it.

I've had thoughts of turning substantial parts of the block cleaner into an airflow-like DAG process that advances through the steps it needs to perform with step retries where possible. Easier to imagine enumerating all possible concurrent executions if it were like that. (Though, it wouldn't address different executions inside of bucketindex methods.)

pracucci commented:
> > > I still haven't figured out how to have a more general test for block cleaning concurrency. Thanks for the fix
> > 
> > 
> > Me neither. I tried reproducing the issue adding random sleeps in the objstore.Bucket functions but I couldn't reproduce it.
> 
> I've had thoughts of turning substantial parts of the block cleaner into an airflow-like DAG process that advances through the steps it needs to perform with step retries where possible. Easier to imagine enumerating all possible concurrent executions if it were like that. (Though, it wouldn't address different executions inside of bucketindex methods.)

Not sure rewriting everything is ever a good idea. The few times we tried we failed spectacularly. I always strongly recommend small iterative changes, and try to do the smallest change to fix an issue when possible (not always possible, but sometimes it is).

pracucci commented:
@dimitarvdimitrov I thought about another possible race condition: the deletion happens between listing deletion markers and fetching their content for **new** markers not existing in the index yet. Due to how deletion delay works it shouldn't happen in the real work (we always expect to delete blocks for which the deletion markers were already in the index), but since the fix is even more trivial I've just fixed it (and added a test case for it too).

dimitarvdimitrov commented on pkg/compactor/blocks_cleaner_test.go:
```diff
@@ -1266,104 +1267,180 @@ func TestBlocksCleaner_RaceCondition_CleanerUpdatesBucketIndexWhileAnotherCleane
+								// because we want cleaner2 to discover the deletion mark before we delete it.
+								<-cleaner2ListMarkersStarted
+
+								// Then signal the deletion happened with some delay. The delay is used to make
+								// sure this deletion completes before the 2nd cleaner is unblocked.
+								once.Do(func() {
+									go func() {
+										time.Sleep(100 * time.Millisecond)
+										close(cleaner2GetDeletionMarkUnblocked)
+									}()
+								})
```
this looks more appropriate for a `postDeleteHook`. It will be less flaky

pracucci replied:
Agree! I don't know why I haven't thought about it before. Done in https://github.com/grafana/mimir/pull/8905/commits/55b097c390f393f939793a330d2d5fd843cfd0b3

pracucci replied:
I've re-run the test on top of `main` and both test cases fail.

dimitarvdimitrov commented on pkg/compactor/blocks_cleaner_test.go:
```diff
@@ -1266,104 +1267,180 @@ func TestBlocksCleaner_RaceCondition_CleanerUpdatesBucketIndexWhileAnotherCleane
+				cleaner1BucketClient = &hookBucket{
+					Bucket: bucketClient,
+					preDeleteHook: func() func(context.Context, string) {
+						once := sync.Once{}
```
Isn't this a new `Once` instance every time the `preDeleteHook` is called? I don't understand how it's different from just calling the func below upon every invocation of `preDeleteHook`. Having said that, I still think a `postDeleteHook` is better.

pracucci replied:
The `Once` is global because a channel can only be closed once, so we protect from it. However, in this specific test we only expect 1 block to be deleted, so it shouldn't matter much.

dimitarvdimitrov replied:
I had missed that it's a `func() func(...)`, so the closure with the once is reused for every call of `Delete()` on the client. SGTM

dimitarvdimitrov reviewed (approved):
Thanks!
